test/remote/routes.js

```diff
@@ -0,0 +1,349 @@
+/* eslint-env mocha */
+'use strict'
+
+const chai = require('chai')
+const dirtyChai = require('dirty-chai')
+const expect = chai.expect
+chai.use(dirtyChai)
+
+const proxyquire = require('proxyquire')
+const multiaddr = require('multiaddr')
+
+const Hapi = require('hapi')
+const routes = proxyquire('../../src/remote-node/routes', {
+  '../daemon-ctrl': class {
+    spawn (ops, cb) {
+      const node = {}
+      node.apiAddr = multiaddr('/ip4/127.0.0.1/tcp/5001')
+      node.gatewayAddr = multiaddr('/ip4/127.0.0.1/tcp/8080')
+      node.started = false
+
+      node.init = (opts, cb) => {
+        cb(null, node)
+      }
+
+      node.cleanup = (cb) => {
+        cb()
+      }
+
+      node.start = (_, cb) => {
+        node.started = true
+
+        const api = {}
+        api.apiHost = node.apiAddr.nodeAddress().address
+        api.apiPort = node.apiAddr.nodeAddress().port
+
+        api.gatewayHost = node.gatewayAddr.nodeAddress().address
+        api.gatewayPort = node.gatewayAddr.nodeAddress().port
+
+        node.api = api
+        cb(null, api)
+      }
+
+      node.stop = (cb) => {
+        node.killProcess(cb)
+      }
+      node.killProcess = (cb) => {
+        node.started = false
+        cb()
+      }
+
+      node.pid = (cb) => {
+        cb(null, 1)
+      }
+
+      node.getConfig = (key, cb) => {
+        cb(null, { foo: 'bar' })
+      }
+
+      node.setConfig = (key, val, cb) => {
+        cb()
+      }
+
+      node.start(null, () => {
+        cb(null, node)
+      })
+    }
+  }
+})
+
+describe('routes', () => {
+  let id
+  const server = new Hapi.Server()
+  before(() => {
+    server.connection()
+    routes(server)
+  })
+
+  after((done) => server.stop(done))
+
+  describe('POST /spawn', () => {
+    it('should return 200', (done) => {
+      server.inject({
+        method: 'POST',
+        url: '/spawn',
+        headers: { 'content-type': 'application/json' }
+      }, (res) => {
+        expect(res.statusCode).to.equal(200)
+        expect(res.result.id).to.exist()
+        expect(res.result.api.apiAddr).to.exist()
+        expect(res.result.api.gatewayAddr).to.exist()
+
+        id = res.result.id
```

Comment:
If I understand it correctly, you define the `id` here for the rest of the tests. Shouldn't the `describes` be independent and be able to be run in any order? Isn't that something for the setup phase of the tests?

Reply:
Sure, we could - but we'd probably end up duplicating `spawn` in a separate test, since `id` is generated by spawn and the rest of the tests wouldn't work without it.